### PR TITLE
Keep track of personal transactions in the database

### PR DESCRIPTION
**Pull request by sampoder**
## Summary of the problem

The whole system of commenting was janky. If we want to make this public we would want to do it better.

## Describe your changes

I made a join table in the database and starting using that to track things.

image

image

### CODE REVIEW

**Comment by sampoder** [comment]
This again is working towards https://github.com/hackclub/hcb/issues/4904

**Comment by sarthaktexas** [comment]
me likey

File: db/schema.rb
Comment (sarthaktexas) [inline db/schema.rb:1]: this is my favorite file in your pr

**Comment by sampoder** [comment]
i'm glad you likey

File: db/schema.rb
Comment (sampoder) [reply db/schema.rb:1]: this is by far my least favourite file.... it's a nightmare

File: app/controllers/hcb_codes_controller.rb [hunk 1]
```diff
@@ -197,7 +197,7 @@ def invoice_as_personal_transaction
 
     return render plain: "404 Not found", status: :not_found if !event&.hack_club_hq? || spender.nil?
 
-    @invoice = ::InvoiceService::Create.new(
+    invoice = ::InvoiceService::Create.new(
```
Comment (garyhtou) [inline app/controllers/hcb_codes_controller.rb]: Move this into a `before_create` callback in `HcbCode::PersonalTransaction`

**Review by garyhtou — COMMENTED** [review]
what are your thoughts on rephrasing it as `Repayment` rather than `Reimbursement`. I don't want people to get it confused with the usual reimbursement.

**Comment by sampoder** [comment]
Repayment is a good word for it!

File: app/models/hcb_code.rb [hunk 2]
```diff
@@ -35,6 +35,8 @@ class HcbCode < ApplicationRecord
   has_many :suggested_pairings
   has_many :suggested_receipts, source: :receipt, through: :suggested_pairings
 
+  has_one :personal_transaction, class_name: "HcbCode::PersonalTransaction", required: false
```
Comment (northeastprince) [inline app/models/hcb_code.rb]:
```suggestion
  has_one :personal_transaction, required: false
```

File: app/models/hcb_code/personal_transaction.rb [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+#  fk_rails_...  (hcb_code_id => hcb_codes.id)
+#  fk_rails_...  (invoice_id => invoices.id)
+#  fk_rails_...  (reviewer_id => users.id)
+#
+class HcbCode
+  class PersonalTransaction < ApplicationRecord
+    belongs_to :hcb_code
+    belongs_to :invoice
+    belongs_to :reviewer
+    self.table_name = "hcb_code_personal_transactions"
```
Comment (northeastprince) [inline app/models/hcb_code/personal_transaction.rb]:
```suggestion
    belongs_to :reviewer
```

File: app/controllers/hcb_codes_controller.rb [hunk 4]
```diff
@@ -191,44 +191,16 @@ def toggle_tag
...
-      current_user:
-    ).run
+    return render plain: "404 Not found", status: :not_found if !event&.hack_club_hq?
 
-    ::HcbCodeService::Comment::Create.new(
-      hcb_code_id: hcb_code.id,
-      content: "This transaction was marked as an accidental misuse. Reimbursement requested at #{hcb_code_url(@invoice.local_hcb_code)}.",
-      current_user:
-    ).run
+    result = HcbCode::PersonalTransaction.create(hcb_code_id: hcb_code.id, reporter_id: current_user.id)
```
Comment (garyhtou) [inline app/controllers/hcb_codes_controller.rb]: No need to pass in the ID, Rails will automatically grab that from the object

```suggestion
    result = HcbCode::PersonalTransaction.create(hcb_code:, reporter: current_user)
```

File: app/controllers/hcb_codes_controller.rb [hunk 5]
```diff
@@ -191,44 +191,16 @@ def toggle_tag
...
-      sponsor_address_postal_code: spender.stripe_cardholder.stripe_billing_address_postal_code,
-      sponsor_address_country: spender.stripe_cardholder.stripe_billing_address_country
-    ).run
-
-    ::HcbCodeService::Comment::Create.new(
-      hcb_code_id: @invoice.local_hcb_code.id,
-      content: "#{hcb_code_url(hcb_code)} was marked as an accidental misuse.",
-      current_user:
-    ).run
+    return render plain: "404 Not found", status: :not_found if !event&.hack_club_hq?
```
Comment (garyhtou) [inline app/controllers/hcb_codes_controller.rb]: Let's move this logic into pundit.

```ruby
class HcbCodePolicy < ApplicationPolicy
  def invoice_as_personal_transaction?
    (user&.admin? || present_in_events?) && record.event&.hack_club_hq?
  end
end
```

File: app/controllers/hcb_codes_controller.rb (hunk 4 above)
Comment (garyhtou) [inline app/controllers/hcb_codes_controller.rb]: can we give `result` another name? maybe `personal_transaction` or `personal_tx`

File: app/models/hcb_code/personal_transaction.rb [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+#  fk_rails_...  (invoice_id => invoices.id)
+#  fk_rails_...  (reviewer_id => users.id)
+#
+class HcbCode
+  class PersonalTransaction < ApplicationRecord
+    belongs_to :hcb_code
+    belongs_to :invoice
+    belongs_to :reporter, class_name: "User"
+
+    before_validation :send_invoice, if: -> { invoice.nil? }
```
Comment (garyhtou) [inline app/models/hcb_code/personal_transaction.rb]:
```suggestion
    after_create :send_invoice, if: -> { invoice.nil? }
```

File: app/models/hcb_code/personal_transaction.rb [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+        sponsor_name: spender.name,
+        sponsor_email: spender.email,
+        sponsor_address_line1: spender.stripe_cardholder.stripe_billing_address_line1,
+        sponsor_address_line2: spender.stripe_cardholder.stripe_billing_address_line2,
+        sponsor_address_city: spender.stripe_cardholder.stripe_billing_address_city,
+        sponsor_address_state: spender.stripe_cardholder.stripe_billing_address_state,
+        sponsor_address_postal_code: spender.stripe_cardholder.stripe_billing_address_postal_code,
+        sponsor_address_country: spender.stripe_cardholder.stripe_billing_address_country
+      ).run
+      self.invoice_id = invoice.id
```
Comment (garyhtou) [inline app/models/hcb_code/personal_transaction.rb]: Directly pass the service's output to `self.invoice=`

```suggestion
      self.invoice = ::InvoiceService::Create.new(
        event_id: event.id,
        due_date: 1.month.from_now,
        item_description: "Reimbursing personal transaction: #{hcb_code.memo}",
        item_amount: hcb_code.amount.abs,
        current_user: reporter,
        sponsor_id: nil,
        sponsor_name: spender.name,
        sponsor_email: spender.email,
        sponsor_address_line1: spender.stripe_cardholder.stripe_billing_address_line1,
        sponsor_address_line2: spender.stripe_cardholder.stripe_billing_address_line2,
        sponsor_address_city: spender.stripe_cardholder.stripe_billing_address_city,
        sponsor_address_state: spender.stripe_cardholder.stripe_billing_address_state,
        sponsor_address_postal_code: spender.stripe_cardholder.stripe_billing_address_postal_code,
        sponsor_address_country: spender.stripe_cardholder.stripe_billing_address_country
      ).run
```

File: app/models/hcb_code/personal_transaction.rb [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+#
+# Foreign Keys
+#
+#  fk_rails_...  (hcb_code_id => hcb_codes.id)
+#  fk_rails_...  (invoice_id => invoices.id)
+#  fk_rails_...  (reviewer_id => users.id)
+#
+class HcbCode
+  class PersonalTransaction < ApplicationRecord
+    belongs_to :hcb_code
```
Comment (garyhtou) [inline app/models/hcb_code/personal_transaction.rb]: Validate `hcb_code` is unique. We only want one `PersonalTransaction` per `HcbCode`

File: app/views/hcb_codes/_accidental_misuse.html.erb [hunk 9]
```diff
@@ -1,4 +1,4 @@
-<% if local_assigns[:hcb_code].event&.hack_club_hq? %>
+<% if local_assigns[:hcb_code].event&.hack_club_hq? && !local_assigns[:hcb_code].personal_transaction.present? %>
```
Comment (garyhtou) [inline app/views/hcb_codes/_accidental_misuse.html.erb]: After moving the `hack_club_hq?` into pundit, we can use that here to conditionally render the partial.

```suggestion
<% if policy(local_asigns[:hcb_code]).invoice_as_personal_transaction? && local_assigns[:hcb_code].personal_transaction.nil? %>
```

https://github.com/varvet/pundit#:~:text=You%20can%20easily%20get%20a%20hold%20of%20an%20instance%20of%20the%20policy%20through%20the%20policy%20method%20in%20both%20the%20view%20and%20controller.%20This%20is%20especially%20useful%20for%20conditionally%20showing%20links%20or%20buttons%20in%20the%20view%3A

**Review by garyhtou — CHANGES_REQUESTED** [review]
Great work! couple minor things

File: app/models/hcb_code/personal_transaction.rb (hunk 6 above)
Comment (sampoder) [reply app/models/hcb_code/personal_transaction.rb]: Making this change breaks things unfortunately: 

Screenshot 2024-01-10 at 2 17 19 PM

Comment (garyhtou) [reply app/models/hcb_code/personal_transaction.rb]: let's try
```suggestion
    before_validation :send_invoice, on: :create, if: -> { invoice.nil? }
```

**Review by garyhtou — APPROVED** [review]

**Pull request merged**